File: ios/LowerBackStretching/Views/StretchAnimation3DView.swift
import SwiftUI

/// 3D-styled mannequin renderer. Same pose data as `StretchAnimationView`;
/// the difference is purely cosmetic:
///  - Each bone is drawn as a thick capsule (wide stroke with rounded caps)
///  - The whole figure is rotated slightly around its vertical axis so it
///    reads as 3/4-perspective rather than flat profile
///  - Bones further from the camera (after the Y-axis rotation) render
///    slightly darker so depth is visible
///  - The head is a radial-gradient sphere with shading
///
/// Pose joints with only `[x, y]` are treated as `z = 0`. Future 3D pose
/// data can supply per-joint depth.
struct StretchAnimation3DView: View {
    let animation: StretchAnimationSpec?
    let youtubeId: String

    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: 12)
                .fill(Color(.secondarySystemBackground))

            if let animation, !animation.poses.isEmpty {
                Mannequin(spec: animation)
            } else {
                // Fallback: 2D placeholder owns the "follow the timer" copy.
                StretchAnimationView(animation: nil, youtubeId: youtubeId)
            }
        }
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }
}

private struct Mannequin: View {
    let spec: StretchAnimationSpec

    var body: some View {
        TimelineView(.animation) { context in
            let secs = context.date.timeIntervalSinceReferenceDate
            let loop = max(spec.loopSeconds, 0.25)
            let phase = secs.truncatingRemainder(dividingBy: loop) / loop
            let t = phase * Double(spec.poses.count)
            let joints = StretchAnimationView.interpolatedPose(poses: spec.poses, t: t)
            Canvas { ctx, size in
                drawMannequin(joints: joints, in: ctx, size: size)
            }
        }
    }
}

private let cameraYRotation: Double = 0.35   // ~20°
private let cameraDistance: Double = 4.0

private struct ProjectedJoint {
    let point: CGPoint
    let depth: Double
    let scale: Double
}

private func drawMannequin(joints: [String: CGPoint], in ctx: GraphicsContext, size: CGSize) {
    if joints.isEmpty { return }
    let w = size.width
    let h = size.height
    let cx = w / 2
    let cy = h / 2
    let minDim = min(w, h)

    var projected: [String: ProjectedJoint] = [:]
    for (name, p) in joints {
        let x = Double(p.x) - 0.5
        let y = -(Double(p.y) - 0.5)
        let z = 0.0
        let rotX = x * cos(cameraYRotation) + z * sin(cameraYRotation)
        let rotZ = -x * sin(cameraYRotation) + z * cos(cameraYRotation)
        let scale = cameraDistance / (cameraDistance + rotZ)
        let px = Double(cx) + rotX * scale * Double(minDim)
        let py = Double(cy) + (-y) * scale * Double(minDim)
        projected[name] = ProjectedJoint(
            point: CGPoint(x: px, y: py),
            depth: rotZ,
            scale: scale
        )
    }

    let baseStroke = max(8.0, Double(minDim) * 0.045)
    let headRadius = Double(minDim) * 0.06

    // Z-sort bones (far first).
    let bones = StretchAnimationView.skeletonBones.compactMap { pair -> (ProjectedJoint, ProjectedJoint, Double)? in
        let (aName, bName) = pair
        guard let pa = projected[aName], let pb = projected[bName] else { return nil }
        return (pa, pb, (pa.depth + pb.depth) / 2.0)
    }.sorted { $0.2 > $1.2 }

    // Same dark-blue shading constants as Android's StretchAnimation3DView
    // and the JS preview, so all three renderers produce visually identical
    // depth cues. The previous iOS implementation reduced opacity for depth,
    // which gave a washed-out look that didn't match Android.
    let bodyColor = Color.accentColor
    let depthDark = Color(red: 27 / 255, green: 58 / 255, blue: 110 / 255)
    let headDark = Color(red: 13 / 255, green: 37 / 255, blue: 69 / 255)

    for (pa, pb, _) in bones {
        let avgScale = (pa.scale + pb.scale) / 2.0
        let stroke = baseStroke * avgScale
        let depthBlend = max(0.0, min(1.0, (pa.depth + pb.depth) / 2.0 + 0.5))

        var path = Path()
        path.move(to: pa.point)
        path.addLine(to: pb.point)

        // Main capsule: blend bone color toward dark-blue with depth.
        let tint = bodyColor.blended(with: depthDark, fraction: depthBlend * 0.5)
        ctx.stroke(
            path,
            with: .color(tint),
            style: StrokeStyle(lineWidth: stroke, lineCap: .round)
        )

        // Highlight stripe: tint blended toward white at 45% alpha.
        let highlight = tint.blended(with: .white, fraction: 0.35).opacity(0.45)
        ctx.stroke(
            path,
            with: .color(highlight),
            style: StrokeStyle(lineWidth: stroke * 0.35, lineCap: .round)
        )
    }

    // Head as a shaded sphere via radial gradient.
    if let head = projected["head"] {
        let r = headRadius * head.scale
        let rect = CGRect(
            x: head.point.x - r,
            y: head.point.y - r,
            width: r * 2,
            height: r * 2
        )
        let gradient = Gradient(colors: [
            bodyColor.blended(with: .white, fraction: 0.4),
            bodyColor,
            bodyColor.blended(with: headDark, fraction: 0.6),
        ])
        ctx.fill(
            Path(ellipseIn: rect),
            with: .radialGradient(
                gradient,
                center: CGPoint(x: head.point.x - r * 0.3, y: head.point.y - r * 0.4),
                startRadius: 0,
                endRadius: r * 1.6
            )
        )
    }
}

// SwiftUI's Color exposes no public component accessor, so blending goes
// through UIColor. This works on all supported iOS versions (no iOS-18-only
// `Color.mix` API needed).
private extension Color {
    func blended(with other: Color, fraction: Double) -> Color {
        let f = CGFloat(max(0.0, min(1.0, fraction)))
        var r1: CGFloat = 0, g1: CGFloat = 0, b1: CGFloat = 0, a1: CGFloat = 0
        var r2: CGFloat = 0, g2: CGFloat = 0, b2: CGFloat = 0, a2: CGFloat = 0
        UIColor(self).getRed(&r1, green: &g1, blue: &b1, alpha: &a1)
        UIColor(other).getRed(&r2, green: &g2, blue: &b2, alpha: &a2)
        return Color(
            red: Double(r1 + (r2 - r1) * f),
            green: Double(g1 + (g2 - g1) * f),
            blue: Double(b1 + (b2 - b1) * f),
            opacity: Double(a1 + (a2 - a1) * f)
        )
    }
}

File: android/app/src/main/java/com/lowerbackstretching/ui/components/StretchAnimation3DView.kt
package com.lowerbackstretching.ui.components

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.unit.dp
import com.lowerbackstretching.core.SKELETON_BONES
import com.lowerbackstretching.core.interpolatedPose
import com.lowerbackstretching.core.model.StretchAnimationSpec
import kotlin.math.cos
import kotlin.math.sin

/**
 * 3D-styled mannequin renderer. Same pose data as [StretchAnimationView];
 * the difference is purely cosmetic:
 *  - Each bone is drawn as a thick capsule (wide stroke + round caps) with
 *    a vertical lightness gradient that fakes a single overhead light
 *  - The head is a sphere drawn as a radial gradient circle (light on top,
 *    dark on bottom)
 *  - The whole figure is rotated slightly around its vertical axis so it
 *    reads as 3/4-perspective rather than flat profile
 *  - Bones further from the camera (after the Y-axis rotation) render with
 *    a slight alpha/desaturation so depth is visible
 *
 * Pose joints with only `[x, y]` are treated as `z = 0`. Future 3D pose
 * data can supply per-joint depth; this renderer projects to 2D at the end.
 */
@Composable
fun StretchAnimation3DView(
    animation: StretchAnimationSpec?,
    youtubeId: String,
    modifier: Modifier = Modifier,
) {
    val surface = MaterialTheme.colorScheme.surfaceVariant
    Box(
        modifier = modifier
            .aspectRatio(16f / 9f)
            .clip(RoundedCornerShape(12.dp))
            .background(surface),
    ) {
        if (animation != null && animation.poses.isNotEmpty()) {
            Mannequin(
                spec = animation,
                bodyColor = MaterialTheme.colorScheme.primary,
                modifier = Modifier.fillMaxSize(),
            )
        } else {
            // Fallback to the 2D placeholder — no animation data to render.
            // The user-facing copy and "Watch demo" link live in the 2D file.
            StretchAnimationView(
                animation = null,
                youtubeId = youtubeId,
                modifier = Modifier.fillMaxSize(),
            )
        }
    }
}

@Composable
private fun Mannequin(
    spec: StretchAnimationSpec,
    bodyColor: Color,
    modifier: Modifier = Modifier,
) {
    val transition = rememberInfiniteTransition(label = "stretch3d")
    val durationMs = (spec.loopSeconds * 1000).toInt().coerceAtLeast(250)
    val t by transition.animateFloat(
        initialValue = 0f,
        targetValue = spec.poses.size.toFloat(),
        animationSpec = infiniteRepeatable(
            animation = tween(durationMillis = durationMs, easing = LinearEasing),
            repeatMode = RepeatMode.Restart,
        ),
        label = "poseT",
    )
    val pose = interpolatedPose(spec.poses, t)
    Canvas(modifier = modifier) {
        drawMannequin(pose, bodyColor)
    }
}

/** How far around the vertical axis to rotate the figure for the 3/4 look,
 *  in radians. Small enough that side-view stretches still read clearly. */
private const val CAMERA_Y_ROTATION_RAD = 0.35f // ~20°

/** Apparent "camera distance" used by the perspective projection. Higher
 *  = less perspective distortion. */
private const val CAMERA_DISTANCE = 4f

private fun DrawScope.drawMannequin(
    joints: Map<String, Pair<Float, Float>>,
    bodyColor: Color,
) {
    if (joints.isEmpty()) return
    val w = size.width
    val h = size.height
    val cx = w / 2f
    val cy = h / 2f

    // Project each joint: normalized [0,1] -> centered NDC [-1,+1] -> rotate
    // around Y axis -> perspective divide -> back to pixel coords.
    val projected = mutableMapOf<String, ProjectedJoint>()
    for ((name, p) in joints) {
        val (nx, ny) = p
        // Center the figure horizontally on its own bounding-box midpoint
        // (we use the canvas center) and convert y into a "up is +" frame
        // so the figure renders right-side-up under rotation.
        val x = (nx - 0.5f)
        val y = -(ny - 0.5f)
        val z = 0f
        val rotX = x * cos(CAMERA_Y_ROTATION_RAD) + z * sin(CAMERA_Y_ROTATION_RAD)
        val rotZ = -x * sin(CAMERA_Y_ROTATION_RAD) + z * cos(CAMERA_Y_ROTATION_RAD)
        val scale = CAMERA_DISTANCE / (CAMERA_DISTANCE + rotZ)
        val px = cx + rotX * scale * size.minDimension
        val py = cy + (-y) * scale * size.minDimension
        projected[name] = ProjectedJoint(Offset(px, py), depth = rotZ, scale = scale)
    }

    val baseStroke = (size.minDimension * 0.045f).coerceAtLeast(8f)
    val headRadius = size.minDimension * 0.06f

    // Z-sort bones so far ones draw first.
    val bones = SKELETON_BONES.mapNotNull { (a, b) ->
        val pa = projected[a] ?: return@mapNotNull null
        val pb = projected[b] ?: return@mapNotNull null
        Triple(pa, pb, (pa.depth + pb.depth) / 2f)
    }.sortedByDescending { it.third }

    for ((pa, pb, _) in bones) {
        val avgScale = (pa.scale + pb.scale) / 2f
        val strokePx = baseStroke * avgScale
        val a = pa.point
        val b = pb.point
        val depthBlend = ((pa.depth + pb.depth) / 2f + 0.5f).coerceIn(0f, 1f)
        val tint = lerpColor(bodyColor, Color(0xFF1B3A6E), depthBlend * 0.5f)
        drawLine(
            color = tint,
            start = a,
            end = b,
            strokeWidth = strokePx,
            cap = StrokeCap.Round,
        )
        val highlight = lerpColor(tint, Color.White, 0.35f)
        drawLine(
            color = highlight.copy(alpha = 0.45f),
            start = a,
            end = b,
            strokeWidth = strokePx * 0.35f,
            cap = StrokeCap.Round,
        )
    }

    // Head — drawn as a radial gradient sphere with shading.
    projected["head"]?.let { h ->
        val r = headRadius * h.scale
        val center = h.point
        val grad = Brush.radialGradient(
            colors = listOf(
                lerpColor(bodyColor, Color.White, 0.4f),
                bodyColor,
                lerpColor(bodyColor, Color(0xFF0D2545), 0.6f),
            ),
            center = Offset(center.x - r * 0.3f, center.y - r * 0.4f),
            radius = r * 1.6f,
        )
        drawCircle(brush = grad, radius = r, center = center)
    }
}

private data class ProjectedJoint(
    val point: Offset,
    val depth: Float,
    val scale: Float,
)

private fun lerpColor(a: Color, b: Color, t: Float): Color {
    val s = t.coerceIn(0f, 1f)
    return Color(
        red = a.red + (b.red - a.red) * s,
        green = a.green + (b.green - a.green) * s,
        blue = a.blue + (b.blue - a.blue) * s,
        alpha = a.alpha + (b.alpha - a.alpha) * s,
    )
}

